File: programacao_dinamica/calcular_anagrama_recursivo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/resource.h>

#define MAX_PALAVRA 100
#define ALPHABET_SIZE 26
#define MAX_FAT 100

typedef struct {
    char *path;
    int qtd_palavras;
    int qtd_testes;
    int mostrar_resultado;
} args;

// =========================== VERSÃO RECURSIVA ===========================

long long fatorialRec(int n, long long *memo) {
    if (n == 0 || n == 1) return 1;
    if (memo && memo[n] != 0) return memo[n];  // proteção contra ponteiro NULL
    long long resultado = n * fatorialRec(n - 1, memo);
    if (memo) memo[n] = resultado;
    return resultado;
}

void contarFrequenciasRec(const char *palavra, int index, int freq[]) {
    if (palavra[index] == '\0') return;
    char c = palavra[index];
    if (c >= 'A' && c <= 'Z') c += 32;
    if (c >= 'a' && c <= 'z') freq[c - 'a']++;
    contarFrequenciasRec(palavra, index + 1, freq);
}

long long calcularDivisorRec(int freq[], int index, long long *memo) {
    if (index == ALPHABET_SIZE) return 1;
    long long atual = 1;
    if (freq[index] > 1) atual = fatorialRec(freq[index], memo);
    return atual * calcularDivisorRec(freq, index + 1, memo);
}

long long calcularAnagramasRecursivo(const char *palavra) {
    long long memo[MAX_FAT] = {0};
    int freq[ALPHABET_SIZE] = {0};
    contarFrequenciasRec(palavra, 0, freq);
    int n = strlen(palavra);
    return fatorialRec(n, memo) / calcularDivisorRec(freq, 0, memo);
}

// =========================== FUNÇÃO PARA MEDIR MEMÓRIA ===========================

long get_mem_kb() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss; // em KB
}

// =========================== ARGUMENTOS E UTILITÁRIOS ===========================

void ler_args(int argc, char *argv[], args *a) {
    if (argc < 5) {
        printf("Uso: %s <arquivo.txt> <qtd_palavras> <qtd_testes> <mostrar_resultado>\n", argv[0]);
        exit(1);
    }
    a->path = argv[1];
    a->qtd_palavras = atoi(argv[2]);
    a->qtd_testes = atoi(argv[3]);
    a->mostrar_resultado = atoi(argv[4]);
    if (a->qtd_palavras <= 0 || a->qtd_testes <= 0) {
        printf("Parâmetros inválidos!\n");
        exit(1);
    }
}

void ler_palavras(char **palavras, int qtd, const char *path) {
    FILE *f = fopen(path, "r");
    if (!f) {
        perror("Erro ao abrir o arquivo");
        exit(1);
    }

    char buffer[MAX_PALAVRA];
    int count = 0;

    while (fgets(buffer, sizeof(buffer), f) && count < qtd) {
        buffer[strcspn(buffer, "\n")] = '\0';
        palavras[count] = malloc(strlen(buffer) + 1);
        strcpy(palavras[count], buffer);
        count++;
    }

    fclose(f);

    if (count < qtd) {
        printf("⚠️  Aviso: arquivo contém apenas %d palavras, mas %d foram solicitadas.\n", count, qtd);
    }
}

// =========================== MAIN ===========================

int main(int argc, char *argv[]) {
    args a;
    ler_args(argc, argv, &a);

    char **palavras = malloc(a.qtd_palavras * sizeof(char *));
    ler_palavras(palavras, a.qtd_palavras, a.path);
    double tempo_rec = 0;
    long long res_rec = 0;
    long memoria_rec = 0;

    for (int i = 0; i < a.qtd_palavras; i++) {
        

        for (int t = 0; t < a.qtd_testes; t++) {
            long mem_before = get_mem_kb();
            clock_t inicio = clock();
            res_rec = calcularAnagramasRecursivo(palavras[i]);
            clock_t fim = clock();
            long mem_after = get_mem_kb();
            tempo_rec += (double)(fim - inicio) / CLOCKS_PER_SEC;
            memoria_rec += (mem_after - mem_before);
        }

        if (a.mostrar_resultado)
            printf("Palavra: %-20s | Rec: %.6fs, %ldKB | Resultado: %lld\n", palavras[i], tempo_rec, memoria_rec,res_rec);
    }
    
    tempo_rec /= a.qtd_testes;
    memoria_rec /= a.qtd_testes;

    printf("Tempo:%.6fs Memoria:%ldKB\n",tempo_rec, memoria_rec);

    for (int i = 0; i < a.qtd_palavras; i++)
        free(palavras[i]);
    free(palavras);

    return 0;
}

File: programacao_dinamica/calcular_anagrama_recursivo.py
import sys
import time
import tracemalloc
import gc
from collections import Counter

class Args:
    def __init__(self, path, qtd_testes, mostrar_resultado):
        self.path = path
        self.qtd_testes = qtd_testes
        self.mostrar_resultado = mostrar_resultado

# =========================== VERSÃO RECURSIVA ===========================

# Função recursiva para calcular fatorial
def fatorial_rec(n):
    if n == 0 or n == 1:
        return 1
    return n * fatorial_rec(n - 1)

# Função recursiva para contar frequências de letras
def contar_frequencias_rec(palavra, index, freq):
    if index >= len(palavra):
        return freq
    c = palavra[index].lower()
    freq[c] += 1  
    return contar_frequencias_rec(palavra, index + 1, freq)

# Função recursiva para calcular o denominador da fórmula
def calcular_denominador_rec(freq_values, index):
    if index >= len(freq_values):
        return 1
    
    count = freq_values[index]
    fat = fatorial_rec(count)
    return fat * calcular_denominador_rec(freq_values, index + 1)

# Função principal para calcular anagramas de forma recursiva
def calcular_anagramas_recursivo(palavra):
    # Inicializa o contador de frequências e o memo para fatoriais
    freq = Counter()
    
    # Conta a frequência de cada letra
    freq = contar_frequencias_rec(palavra, 0, freq)
    
    # Calcula o numerador (fatorial do tamanho da palavra)
    numerador = fatorial_rec(len(palavra))
    
    # Calcula o denominador (produto dos fatoriais das frequências)
    denominador = calcular_denominador_rec(list(freq.values()), 0)
    
    # Retorna o resultado da fórmula
    return numerador // denominador

# =========================== FUNÇÃO PARA MEDIR MEMÓRIA E TEMPO ===========================

def measure_performance(func, *args):
    gc.collect()  # Força coleta de lixo antes da medição
    
    # Mede tempo
    inicio = time.time()
    
    # Mede memória
    tracemalloc.start()
    result = func(*args)
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    
    # Calcula tempo
    fim = time.time()
    tempo = fim - inicio
    
    return result, tempo, peak / 1024  # Resultado, tempo em segundos, memória em KB

# =========================== ARGUMENTOS E UTILITÁRIOS ===========================

def ler_args(argc, argv):
    if argc < 4:
        print(f"Uso: {argv[0]} <arquivo.txt> <qtd_testes> <mostrar_resultado>")
        sys.exit(1)
    
    path = argv[1]
    qtd_testes = int(argv[2])
    mostrar_resultado = int(argv[3])
    return Args(path, qtd_testes, mostrar_resultado)

def ler_palavra(path):
    try:
        with open(path, 'r') as f:
            palavra = f.readline().strip()
            return palavra
    except FileNotFoundError:
        print(f"Erro: arquivo {path} não encontrado.")
        sys.exit(1)

# =========================== MAIN ===========================

def main():
    args = ler_args(len(sys.argv), sys.argv)
    palavra = ler_palavra(args.path)
    
    tempo_total = 0
    memoria_total = 0
    resultado = None
    
    for i in range(args.qtd_testes):
        res, tempo, memoria = measure_performance(calcular_anagramas_recursivo, palavra)
        tempo_total += tempo
        memoria_total += memoria
        resultado = res
    
    tempo_medio = tempo_total / args.qtd_testes
    memoria_media = memoria_total / args.qtd_testes
    
    if args.mostrar_resultado:
        print(f"Palavra: {palavra}")
        print(f"Quantidade de anagramas: {resultado}")
        print(f"Tempo médio: {tempo_medio:.6f}s")
        print(f"Memória média: {memoria_media:.2f} KB")
    
    print(f"Tempo:{tempo_medio:.6f}s Memoria:{memoria_media:.2f}KB QTD:{len(palavra)}")

if __name__ == "__main__":
    main()